pkg/istio-agent/xds_proxy.go

```diff
@@ -708,3 +730,85 @@ func sendWithTimeout(ctx context.Context, sendFunc func(errorChan chan error)) e
 		return err
 	}
 }
+
+// tapRequest() sends "req" to Istiod, and returns a matching response, or `nil` on timeout.
+// Requests are serialized -- only one may be in-flight at a time.
+func (p *XdsProxy) tapRequest(req *discovery.DiscoveryRequest, timeout time.Duration) (*discovery.DiscoveryResponse, error) {
+	if p.connected == nil {
+		return nil, fmt.Errorf("proxy not connected to Istiod")
+	}
+
+	// Only allow one tap request at a time
+	p.tapMutex.Lock()
+	defer p.tapMutex.Unlock()
+
+	// Send to Istiod
+	p.connected.requestsChan <- req
+
+	// Wait for expected response or timeout
+	for {
+		select {
+		case res := <-p.tapResponseChannel:
+			if res.TypeUrl == req.TypeUrl {
+				return res, nil
+			}
+		case <-time.After(timeout):
+			return nil, nil
+		}
+	}
+}
+
+func (p *XdsProxy) makeTapHandler() func(w http.ResponseWriter, req *http.Request) {
+	return func(w http.ResponseWriter, req *http.Request) {
+		typeURL := fmt.Sprintf("istio.io%s", req.URL.Path)
+		response, err := p.tapRequest(&discovery.DiscoveryRequest{
+			TypeUrl: typeURL,
+		}, 5*time.Second)
+		if err != nil {
+			w.WriteHeader(503)
+			fmt.Fprintf(w, "%v\n", err)
+			return
+		}
+
+		if response == nil {
+			log.Infof("timed out waiting for Istiod to respond to %q", typeURL)
+			w.WriteHeader(504)
+			return
+		}
+
+		// Send Istiod's response as HTTP body
+		jsonm := &jsonpb.Marshaler{Indent: "  "}
+		w.Header().Add("Content-Type", "application/json")
+		_ = jsonm.Marshal(w, response)
```

Comment:
nit: Can the line "_ = jsonm.Marshal(w, response)" be simplified to "jsonm.Marshal(w, response)"?

Reply:
That's there for `lint`.  If I don't do it, Istio's build process says **Error return value of `jsonm.Marshal` is not checked**.

It would be possible to Marshal to a buffer, check the result, and then send that result to `w`.  It seemed like a lot of work, given that it isn't really possible for this code to fail.